src/main/scala/com/fulcrumgenomics/bam/ExtractUmisFromBam.scala

```diff
@@ -0,0 +1,154 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.{MolecularBarcode, ReadSegment, ReadStructure, Template}
+import dagr.commons.CommonsDef.PathToBam
+import dagr.commons.io.Io
+import dagr.commons.util.LazyLogging
+import dagr.sopt._
+import dagr.sopt.cmdline.ValidationException
+import htsjdk.samtools._
+import htsjdk.samtools.util._
+
+import scala.collection.JavaConversions._
+
+@clp(description =
+  """
+    |Extracts unique molecular indices from reads in a BAM file into tags.
+    |
+    |Currently only paired end read data are supported.
+    |
+    |Only template bases will be retained as read bases (stored in the SEQ field) as specified by the read structure.
+    |
+    |For a given read, if only one molecular barcode tag is given, then the molecular barcodes will be concatenated and
+    |stored in that tag.  Otherwise the number of molecular barcodes should match the number of tags.  Each  end of a
+    |pair will store the molecular barcode(s) for both pairs.  In this case, the tags for the molecular barcodes should
+    |be unique across both ends. Optionally, the read names can be annotated with the molecular barcode(s) directly.
+    |In this case, the read name will be formatted "<NAME>|<UMIs1><UMIs2>" where "<UMIs1>" is the concatenation of read
+    |one's molecular barcodes.  Similarly for "<UMIs2>".
+    |
+    |Mapping information will be unchanged, as such, this tool should not be used on reads that have been mapped since
+    |it will lead to an BAM with inconsistent records.
+  """,
+  group = ClpGroups.SamOrBam)
+class ExtractUmisFromBam
+( @arg(flag = "i", doc = "Input BAM file.")                                      val input: PathToBam,
+  @arg(flag = "o", doc = "Output BAM file.")                                     val output: PathToBam,
+  @arg(flag = "r", doc = "The read structure for read 1.")                       val readStructure1: String,
+  @arg(flag = "R", doc = "The read structure for read 2.")                       val readStructure2: String,
+  @arg(flag = "b", doc = "SAM tags in which to store the molecular barcodes (one-per segment) for read 1.")
+                                                                                 val molecularBarcodeTags1: Seq[String] = Seq.empty,
+  @arg(flag = "B", doc = "SAM tags in which to store the molecular barcodes (one-per segment) for read 2.")
```

Comment:
If it's not symmetric, which we have seen, then I don't like `ExtractUmisFromBam ... -r 8M8M75T -R 8M75T -b ZA ZB ZC` since the read structure was in two args but the tags are not.  I (personally) like: 
`ExtractUmisFromBam ... -r 8M8M75T -R 8M75T -b ZA ZB -B ZC`.